app/update/update_util.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+define('PACKAGE_PATHNAME', DATA_PATH . '/update_package');
+
+
+// Delete a file tree (from http://php.net/rmdir#110489)
+function del_tree($dir) {
+	if (!is_dir($dir)) {
+		return true;
+	}
+
+	$files = array_diff(scandir($dir), array('.', '..'));
+	foreach ($files as $filename) {
+		$filename = $dir . '/' . $filename;
+		if (is_dir($filename)) {
+			@chmod($filename, 0777);
+			del_tree($filename);
+		} else {
+			unlink($filename);
+		}
+	}
+
+	return rmdir($dir);
+}
+
+
+// Do a recursive copy (from http://fr2.php.net/manual/en/function.copy.php#91010)
+function recurse_copy($src, $dst) {
+	$dir = opendir($src);
+	@mkdir($dst);
+	while (false !== ($file = readdir($dir))) {
+		if (($file != '.') && ($file != '..')) {
+			if (is_dir($src . '/' . $file)) {
+				recurse_copy($src . '/' . $file, $dst . '/' . $file);
+			} else {
+				copy($src . '/' . $file, $dst . '/' . $file);
+			}
+		}
+	}
+	closedir($dir);
+
+	return true;
+}
+
+
+// Remove previous backup of data.
+function remove_data_backup() {
+	return del_tree(DATA_PATH . '.back');
+}
+
+
+// Create a backup of data.
+function data_backup() {
+	$from = DATA_PATH;
+	$to = $from .'.back';
+
+	if (file_exists($from)) {
+		return recurse_copy($from, $to);
+	}
+
+	return false;
+}
+
+
+// Check if a directory exists and if it is writable.
+function check_directory($dir) {
+	if (!file_exists($dir)) {
+		$res = @mkdir($dir, 0775);
+		if (!$res) {
+			return false;
+		}
+	}
+
+	if (!is_writable($dir)) {
+		$res = chmod($dir, 0775);
+		if (!$res) {
+			return false;
+		}
+	}
+
+	$index = $dir . '/index.html';
+	$data = <<<EOF
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en-GB" lang="en-GB">
+<head>
+<meta charset="UTF-8" />
+<meta http-equiv="Refresh" content="0; url=/" />
```

Comment:
Is this root `/` quite what is wanted here?

Reply:
This is some code from Marien, directly from the current update server.
This part is generating a index.html. I have a habit of putting an index.html file in directories that should not be browsed (even when there are other rules such as .htaccess, etc.). The redirect address is not very important, as it is not supposed to be used (except during manual / accidental browsing).